probe-rs/src/session.rs

```diff
@@ -1,14 +1,121 @@
-use crate::probe::MasterProbe;
+use crate::probe::{DebugProbeError, MasterProbe};
 
 pub struct Session {
     pub target: Target,
     pub probe: MasterProbe,
+
+    hw_breakpoint_enabled: bool,
+    active_breakpoints: Vec<Breakpoint>,
+    bp_id_count: usize,
 }
 
 impl Session {
     /// Open a new session with a given debug target
     pub fn new(target: Target, probe: MasterProbe) -> Self {
-        Self { target, probe }
+        Self {
+            target,
+            probe,
+            hw_breakpoint_enabled: false,
+            active_breakpoints: Vec::new(),
+            bp_id_count: 0,
+        }
+    }
+
+    /// Set a hardware breakpoint
+    pub fn set_hw_breakpoint(&mut self, address: u32) -> Result<BreakpointId, DebugProbeError> {
+        log::debug!("Trying to set HW breakpoint at address {:#08x}", address);
+
+        // Get the number of HW breakpoints available
+        let num_hw_breakpoints =
+            self.target
+                .core
+                .get_available_breakpoint_units(&mut self.probe)? as usize;
+
+        log::debug!("{} HW breakpoints are supported.", num_hw_breakpoints);
+
+        if num_hw_breakpoints <= self.active_breakpoints.len() {
+            // We cannot set additional breakpoints
+            log::warn!("Maximum number of breakpoints ({}) reached, unable to set additional HW breakpoint.", num_hw_breakpoints);
+
+            // TODO: Better error here
+            return Err(DebugProbeError::UnknownError);
+        }
+
+        if !self.hw_breakpoint_enabled {
+            self.target.core.enable_breakpoints(&mut self.probe, true)?;
+            self.hw_breakpoint_enabled = true;
+        }
+
+        let bp_unit = self.find_free_breakpoint_unit();
+
+        log::debug!("Using comparator {} of breakpoint unit", bp_unit);
+        // actually set the breakpoint
+        self.target
+            .core
+            .set_breakpoint(&mut self.probe, bp_unit, address)?;
+
+        let id = BreakpointId(self.bp_id_count);
+        self.bp_id_count += 1;
+
+        self.active_breakpoints.push(Breakpoint {
+            id,
+            register_hw: bp_unit,
+        });
+
+        Ok(id)
+    }
+
+    pub fn clear_hw_breakpoint(&mut self, id: BreakpointId) -> Result<(), DebugProbeError> {
+        let bp_position = self.active_breakpoints.iter().position(|bp| bp.id == id);
+
+        match bp_position {
+            Some(bp_position) => {
+                let bp = &self.active_breakpoints[bp_position];
+                self.target
+                    .core
+                    .clear_breakpoint(&mut self.probe, bp.register_hw)?;
+
+                // We only remove the breakpoint if we have actually managed to clear it.
+                self.active_breakpoints.swap_remove(bp_position);
+                Ok(())
+            }
+            None => Err(DebugProbeError::UnknownError),
+        }
+    }
+
+    fn find_free_breakpoint_unit(&self) -> usize {
+        let mut used_bp: Vec<_> = self
```

Comment:
Basically same here with the HashMap vs Vec. Maybe I am overlooking somethin, but this seems more straightforward with a HashMap.

Reply:
I don't see how the hash map tells me which register in the breakpoint unit is free. Do you have an idea how it could be solved more easy with a hash map?

Reply:
Agreed it doesn't help much here, dunno what I thought. Maybe I just wanted to strenghtn my point for it :/